File: core/contracts/funds.py
"""Fund intelligence domain contracts."""

from datetime import date
from typing import Dict, List, Optional

from pydantic import BaseModel, Field, field_validator


class FundMaster(BaseModel):
    """基金主数据。"""

    symbol: str = Field(description="Fund symbol, such as 000001.OF")
    name: str = Field(description="Fund name")
    fund_type: Optional[str] = Field(default=None, description="Fund category")
    management_company: Optional[str] = Field(default=None, description="Management company")
    inception_date: Optional[date] = Field(default=None, description="Fund inception date")
    benchmark: Optional[str] = Field(default=None, description="Performance benchmark")
    latest_size: Optional[float] = Field(default=None, description="Latest fund size")

    @field_validator("symbol", "name")
    @classmethod
    def _not_blank(cls, value: str) -> str:
        if not value or not value.strip():
            raise ValueError("value must not be blank")
        return value.strip()


class FundNavPoint(BaseModel):
    """基金日净值数据点。"""

    symbol: str = Field(description="Fund symbol")
    trading_day: date = Field(description="Trading day")
    unit_nav: float = Field(description="Unit NAV")
    accumulated_nav: Optional[float] = Field(default=None, description="Accumulated NAV")
    daily_return: Optional[float] = Field(default=None, description="Daily return")


class FundHolding(BaseModel):
    """基金股票持仓。"""

    symbol: str = Field(description="Fund symbol")
    report_date: date = Field(description="Disclosure report date")
    stock_symbol: str = Field(description="Held stock symbol")
    stock_name: Optional[str] = Field(default=None, description="Held stock name")
    industry: Optional[str] = Field(default=None, description="Industry label")
    theme: Optional[str] = Field(default=None, description="Theme label")
    weight: float = Field(description="Holding weight in fund NAV")
    market_value: Optional[float] = Field(default=None, description="Holding market value")


class FundManagerProfile(BaseModel):
    """基金经理画像。"""

    manager_id: str = Field(description="Manager identifier")
    manager_name: str = Field(description="Manager name")
    institution_name: Optional[str] = Field(default=None, description="Institution name")
    tenure_start: Optional[date] = Field(default=None, description="Tenure start date")
    tenure_end: Optional[date] = Field(default=None, description="Tenure end date")


class FundPerformanceMetrics(BaseModel):
    """基金收益风险指标。"""

    total_return: Optional[float] = Field(default=None, description="Total return")
    annualized_return: Optional[float] = Field(default=None, description="Annualized return")
    max_drawdown: Optional[float] = Field(default=None, description="Maximum drawdown")
    volatility: Optional[float] = Field(default=None, description="Annualized volatility")
    win_rate: Optional[float] = Field(default=None, description="Positive-return day ratio")
    ulcer_index: Optional[float] = Field(default=None, description="Ulcer index")


class FundExposureBreakdown(BaseModel):
    """基金或组合穿透后的暴露项。"""

    label: str = Field(description="Exposure label")
    weight: float = Field(description="Exposure weight")


class FundDetail(BaseModel):
    """基金详情聚合对象。"""

    master: FundMaster = Field(description="Fund master data")
    managers: List[FundManagerProfile] = Field(
        default_factory=list, description="Current or historical managers"
    )
    latest_nav: Optional[FundNavPoint] = Field(default=None, description="Latest NAV point")
    performance: FundPerformanceMetrics = Field(
        default_factory=FundPerformanceMetrics, description="Performance metrics"
    )
    latest_holdings: List[FundHolding] = Field(
        default_factory=list, description="Latest disclosed holdings"
    )


class PortfolioFundExposure(BaseModel):
    """基金组合穿透结果。"""

    positions: Dict[str, float] = Field(description="Fund symbol to portfolio weight")
    industry_exposure: List[FundExposureBreakdown] = Field(default_factory=list)
    stock_exposure: List[FundExposureBreakdown] = Field(default_factory=list)
    theme_exposure: List[FundExposureBreakdown] = Field(default_factory=list)

File: app/web/static/js/funds.js
/* ============================================================
   AlphaFoundry — Fund Intelligence Panel
   ============================================================ */

import { apiCall, toast, esc } from './core.js';

let initialized = false;

function setText(id, value) {
    const el = document.getElementById(id);
    if (el) el.textContent = value ?? '--';
}

function setStatus(text, kind = 'warning') {
    const badge = document.getElementById('fund-status-badge');
    if (!badge) return;
    badge.textContent = text;
    badge.className = `badge badge-${kind}`;
}

function fmtNumber(value, digits = 4) {
    if (value === undefined || value === null || Number.isNaN(Number(value))) return '--';
    return Number(value).toLocaleString('zh-CN', {
        minimumFractionDigits: 0,
        maximumFractionDigits: digits,
    });
}

function fmtPercent(value) {
    if (value === undefined || value === null || Number.isNaN(Number(value))) return '--';
    return `${(Number(value) * 100).toFixed(2)}%`;
}

function normalizeExposureRows(rows) {
    return Array.isArray(rows) ? rows.filter(row => row && row.label) : [];
}

function renderExposureList(containerId, rows) {
    const container = document.getElementById(containerId);
    if (!container) return;

    const normalized = normalizeExposureRows(rows).slice(0, 12);
    if (!normalized.length) {
        container.innerHTML = '<p class="empty-state compact">暂无数据</p>';
        return;
    }

    container.innerHTML = normalized.map(row => {
        const weight = Number(row.weight || 0);
        const width = Math.max(2, Math.min(100, Math.abs(weight) * 100));
        return `
            <div class="fund-exposure-row">
                <span class="fund-exposure-label">${esc(row.label)}</span>
                <span class="fund-exposure-bar"><i style="width:${width}%"></i></span>
                <strong>${fmtPercent(weight)}</strong>
            </div>`;
    }).join('');
}

function renderManagers(managers) {
    const container = document.getElementById('fund-managers');
    if (!container) return;

    if (!Array.isArray(managers) || !managers.length) {
        container.innerHTML = '<p class="empty-state compact">暂无数据</p>';
        return;
    }

    container.innerHTML = managers.map(manager => `
        <div class="fund-list-item">
            <strong>${esc(manager.manager_name || manager.manager_id)}</strong>
            <span>${esc(manager.institution_name || '--')}</span>
            <small>${esc(manager.tenure_start || '--')} 至 ${esc(manager.tenure_end || '今')}</small>
        </div>
    `).join('');
}

function renderHoldings(holdings) {
    const container = document.getElementById('fund-holdings-table');
    if (!container) return;

    if (!Array.isArray(holdings) || !holdings.length) {
        container.innerHTML = '<p class="empty-state compact">暂无数据</p>';
        return;
    }

    container.innerHTML = `
        <table class="compact">
            <thead>
                <tr>
                    <th>股票代码</th>
                    <th>名称</th>
                    <th>行业</th>
                    <th>主题</th>
                    <th>权重</th>
                    <th>市值</th>
                    <th>报告期</th>
                </tr>
            </thead>
            <tbody>
                ${holdings.map(row => `
                    <tr>
                        <td>${esc(row.stock_symbol)}</td>
                        <td>${esc(row.stock_name || '--')}</td>
                        <td>${esc(row.industry || '--')}</td>
                        <td>${esc(row.theme || '--')}</td>
                        <td>${fmtPercent(row.weight)}</td>
                        <td>${fmtNumber(row.market_value, 2)}</td>
                        <td>${esc(row.report_date || '--')}</td>
                    </tr>
                `).join('')}
            </tbody>
        </table>`;
}

function renderFundDetail(detail) {
    const master = detail.master || {};
    const latestNav = detail.latest_nav || {};
    const performance = detail.performance || {};

    setText('fund-name', master.name || '--');
    setText('fund-symbol-display', master.symbol || '--');
    setText('fund-type', master.fund_type || '类型 --');
    setText('fund-company', master.management_company || '管理人 --');
    setText('fund-benchmark', master.benchmark || '基准 --');
    setText('fund-unit-nav', fmtNumber(latestNav.unit_nav));
    setText('fund-acc-nav', fmtNumber(latestNav.accumulated_nav));
    setText('fund-total-return', fmtPercent(performance.total_return));
    setText('fund-max-drawdown', fmtPercent(performance.max_drawdown));
    setText('fund-volatility', fmtPercent(performance.volatility));
    setText('fund-win-rate', fmtPercent(performance.win_rate));
    renderManagers(detail.managers);
    renderHoldings(detail.latest_holdings);
}

async function loadFund() {
    const input = document.getElementById('fund-symbol');
    const loading = document.getElementById('fund-loading');
    const result = document.getElementById('fund-result');
    const symbol = input?.value?.trim();

    if (!symbol) {
        toast('请输入基金代码', 'error');
        return;
    }

    loading?.classList.remove('hidden');
    result?.classList.add('hidden');
    setStatus('查询中...', 'warning');

    try {
        const detail = await apiCall('GET', `/api/funds/${encodeURIComponent(symbol)}`);
        const exposure = await apiCall('GET', `/api/funds/${encodeURIComponent(symbol)}/exposure`);
        renderFundDetail(detail);
        renderExposureList('fund-industry-exposure', exposure.industry_exposure);
        result?.classList.remove('hidden');
        setStatus('已加载', 'success');
        toast('基金情报已更新', 'success');
    } catch (error) {
        console.error('[funds] load fund failed', error);
        setStatus('查询失败', 'error');
        toast(`基金查询失败: ${error.message || error}`, 'error');
    } finally {
        loading?.classList.add('hidden');
    }
}

function parsePortfolioPositions(raw) {
    const parsed = JSON.parse(raw || '{}');
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
        throw new Error('组合输入必须是对象');
    }

    return Object.fromEntries(
        Object.entries(parsed)
            .filter(([symbol]) => symbol && symbol.trim())
            .map(([symbol, weight]) => [symbol.trim(), Number(weight)])
            .filter(([, weight]) => Number.isFinite(weight))
    );
}

async function calculatePortfolioExposure() {
    const input = document.getElementById('fund-portfolio-input');
    const container = document.getElementById('fund-portfolio-result');

    try {
        const positions = parsePortfolioPositions(input?.value);
        if (!Object.keys(positions).length) {
            toast('请输入有效组合权重', 'error');
            return;
        }

        const exposure = await apiCall('POST', '/api/funds/portfolio/exposure', { positions });
        if (container) {
            container.innerHTML = `
                <div class="fund-exposure-group">
                    <h4>行业</h4>
                    <div id="fund-portfolio-industry"></div>
                </div>
                <div class="fund-exposure-group">
                    <h4>股票</h4>
                    <div id="fund-portfolio-stock"></div>
                </div>
                <div class="fund-exposure-group">
                    <h4>主题</h4>
                    <div id="fund-portfolio-theme"></div>
                </div>`;
        }
        renderExposureList('fund-portfolio-industry', exposure.industry_exposure);
        renderExposureList('fund-portfolio-stock', exposure.stock_exposure);
        renderExposureList('fund-portfolio-theme', exposure.theme_exposure);
        toast('组合穿透已更新', 'success');
    } catch (error) {
        console.error('[funds] calculate portfolio exposure failed', error);
        toast(`组合穿透失败: ${error.message || error}`, 'error');
    }
}

function parseRows(raw) {
    const rows = JSON.parse(raw || '[]');
    if (!Array.isArray(rows)) {
        throw new Error('Rows JSON 必须是数组');
    }
    return rows;
}

async function ingestFundRows() {
    const dataset = document.getElementById('fund-ingest-dataset')?.value || 'master';
    const source = document.getElementById('fund-ingest-source')?.value?.trim() || 'web';
    const rowsRaw = document.getElementById('fund-ingest-rows')?.value;
    const result = document.getElementById('fund-ingest-result');

    try {
        const rows = parseRows(rowsRaw);
        if (!rows.length) {
            toast('Rows JSON 为空', 'error');
            return;
        }
        const response = await apiCall('POST', '/api/funds/ingest', { dataset, source, rows });
        if (result) {
            result.innerHTML = `
                <span>dataset: <strong>${esc(response.dataset || dataset)}</strong></span>
                <span>fetched: <strong>${esc(String(response.fetched ?? 0))}</strong></span>
                <span>saved: <strong>${esc(String(response.saved ?? 0))}</strong></span>
            `;
        }
        toast('基金数据已导入', 'success');
    } catch (error) {
        console.error('[funds] ingest rows failed', error);
        toast(`基金数据导入失败: ${error.message || error}`, 'error');
    }
}

export function initFundsPanel() {
    if (initialized) return;
    initialized = true;

    document.getElementById('btn-load-fund')?.addEventListener('click', loadFund);
    document.getElementById('fund-symbol')?.addEventListener('keydown', event => {
        if (event.key === 'Enter') loadFund();
    });
    document.getElementById('btn-calc-fund-portfolio')?.addEventListener('click', calculatePortfolioExposure);
    document.getElementById('btn-ingest-fund-rows')?.addEventListener('click', ingestFundRows);
}

window.initFundsPanel = initFundsPanel;
